File: core/impl/port/backend/codex/schema_gate.go
// Package codex implements the codex backend adapter (core/domain/port/backend).
package codex

import "sort"

// ApplyOpenAIStrictSchemaGate mutates body in place, normalizing the
// structured-output text.format.schema and every tool parameter schema to
// satisfy the OpenAI Responses API's strict json_schema mode.
func ApplyOpenAIStrictSchemaGate(body map[string]any) {
	normalizeTextFormatSchema(body)
	normalizeToolParameterSchemas(body)
}

// NormalizeOpenAIStrictResponseSchema returns a normalized copy of schema.
func NormalizeOpenAIStrictResponseSchema(schema map[string]any) map[string]any {
	out := deepCopyMap(schema)
	normalizeOpenAIStrictSchemaValue(out)
	return out
}

func normalizeTextFormatSchema(body map[string]any) {
	text, ok := body["text"].(map[string]any)
	if !ok {
		return
	}
	format, ok := text["format"].(map[string]any)
	if !ok {
		return
	}
	schema, ok := format["schema"].(map[string]any)
	if !ok {
		return
	}
	normalizeOpenAIStrictSchemaValue(schema)
}

func normalizeToolParameterSchemas(body map[string]any) {
	tools, ok := body["tools"].([]any)
	if !ok {
		return
	}
	for _, t := range tools {
		tool, ok := t.(map[string]any)
		if !ok {
			continue
		}
		if parameters, ok := tool["parameters"].(map[string]any); ok {
			normalizeOpenAIStrictSchemaValue(parameters)
		}
		if function, ok := tool["function"].(map[string]any); ok {
			if parameters, ok := function["parameters"].(map[string]any); ok {
				normalizeOpenAIStrictSchemaValue(parameters)
			}
		}
	}
}

func normalizeOpenAIStrictSchemaValue(obj map[string]any) {
	if isObjectSchema(obj) {
		normalizeOpenAIStrictObjectSchema(obj)
	}

	for _, key := range []string{"items", "additionalProperties", "contains", "not", "if", "then", "else"} {
		if value, ok := obj[key].(map[string]any); ok {
			normalizeOpenAIStrictSchemaValue(value)
		}
	}

	for _, key := range []string{"anyOf", "oneOf", "allOf"} {
		if values, ok := obj[key].([]any); ok {
			for _, v := range values {
				if m, ok := v.(map[string]any); ok {
					normalizeOpenAIStrictSchemaValue(m)
				}
			}
		}
	}

	for _, key := range []string{"$defs", "definitions"} {
		if defs, ok := obj[key].(map[string]any); ok {
			for _, v := range defs {
				if m, ok := v.(map[string]any); ok {
					normalizeOpenAIStrictSchemaValue(m)
				}
			}
		}
	}
}

func normalizeOpenAIStrictObjectSchema(obj map[string]any) {
	obj["additionalProperties"] = false

	originalRequired := map[string]bool{}
	if required, ok := obj["required"].([]any); ok {
		for _, r := range required {
			if s, ok := r.(string); ok {
				originalRequired[s] = true
			}
		}
	}

	if _, exists := obj["properties"]; !exists {
		obj["properties"] = map[string]any{}
	}
	properties, ok := obj["properties"].(map[string]any)
	if !ok {
		return
	}

	propertyNames := make([]string, 0, len(properties))
	for name := range properties {
		propertyNames = append(propertyNames, name)
	}
	sort.Strings(propertyNames)

	for _, name := range propertyNames {
		property, ok := properties[name].(map[string]any)
		if !ok {
			continue
		}
		normalizeOpenAIStrictSchemaValue(property)
		if !originalRequired[name] {
			makeSchemaNullable(properties, name, property)
		}
	}

	required := make([]any, len(propertyNames))
	for i, name := range propertyNames {
		required[i] = name
	}
	obj["required"] = required
}

func isObjectSchema(obj map[string]any) bool {
	if _, ok := obj["properties"]; ok {
		return true
	}
	ty, ok := obj["type"].(string)
	return ok && ty == "object"
}

func makeSchemaNullable(properties map[string]any, name string, schema map[string]any) {
	typeValue, hasType := schema["type"]
	if !hasType {
		properties[name] = map[string]any{
			"anyOf": []any{
				schema,
				map[string]any{"type": "null"},
			},
		}
		return
	}

	switch ty := typeValue.(type) {
	case string:
		if ty != "null" {
			schema["type"] = []any{ty, "null"}
		}
	case []any:
		for _, v := range ty {
			if s, ok := v.(string); ok && s == "null" {
				return
			}
		}
		schema["type"] = append(ty, "null")
	default:
		// type present but neither string nor array: leave unchanged,
		// mirrors the Rust Some(_) => {} branch (schema_gate.rs:167).
	}
}

func deepCopyMap(in map[string]any) map[string]any {
	out := make(map[string]any, len(in))
	for k, v := range in {
		out[k] = deepCopyValue(v)
	}
	return out
}

func deepCopyValue(v any) any {
	switch value := v.(type) {
	case map[string]any:
		return deepCopyMap(value)
	case []any:
		out := make([]any, len(value))
		for i, e := range value {
			out[i] = deepCopyValue(e)
		}
		return out
	default:
		return value
	}
}

File: old_rust/crates/gateway-backend-codex/src/schema_gate.rs
#![forbid(unsafe_code)]

pub fn apply_openai_strict_schema_gate(body: &mut serde_json::Value) {
    normalize_text_format_schema(body);
    normalize_tool_parameter_schemas(body);
}

#[must_use]
pub fn normalize_openai_strict_response_schema(schema: &serde_json::Value) -> serde_json::Value {
    let mut schema = schema.clone();
    normalize_openai_strict_schema_value(&mut schema);
    schema
}

fn normalize_text_format_schema(body: &mut serde_json::Value) {
    let Some(schema) = body
        .get_mut("text")
        .and_then(|text| text.get_mut("format"))
        .and_then(|format| format.get_mut("schema"))
    else {
        return;
    };

    normalize_openai_strict_schema_value(schema);
}

fn normalize_tool_parameter_schemas(body: &mut serde_json::Value) {
    let Some(tools) = body
        .get_mut("tools")
        .and_then(serde_json::Value::as_array_mut)
    else {
        return;
    };

    for tool in tools {
        if let Some(parameters) = tool.get_mut("parameters") {
            normalize_openai_strict_schema_value(parameters);
        }

        if let Some(parameters) = tool
            .get_mut("function")
            .and_then(|function| function.get_mut("parameters"))
        {
            normalize_openai_strict_schema_value(parameters);
        }
    }
}

fn normalize_openai_strict_schema_value(schema: &mut serde_json::Value) {
    let Some(obj) = schema.as_object_mut() else {
        return;
    };

    if is_object_schema(obj) {
        normalize_openai_strict_object_schema(obj);
    }

    for key in [
        "items",
        "additionalProperties",
        "contains",
        "not",
        "if",
        "then",
        "else",
    ] {
        if let Some(value) = obj.get_mut(key) {
            normalize_openai_strict_schema_value(value);
        }
    }

    for key in ["anyOf", "oneOf", "allOf"] {
        if let Some(values) = obj.get_mut(key).and_then(serde_json::Value::as_array_mut) {
            for value in values {
                normalize_openai_strict_schema_value(value);
            }
        }
    }

    for key in ["$defs", "definitions"] {
        if let Some(defs) = obj.get_mut(key).and_then(serde_json::Value::as_object_mut) {
            for value in defs.values_mut() {
                normalize_openai_strict_schema_value(value);
            }
        }
    }
}

fn normalize_openai_strict_object_schema(obj: &mut serde_json::Map<String, serde_json::Value>) {
    obj.insert(
        "additionalProperties".to_string(),
        serde_json::Value::Bool(false),
    );

    let original_required = obj
        .get("required")
        .and_then(serde_json::Value::as_array)
        .map(|values| {
            values
                .iter()
                .filter_map(serde_json::Value::as_str)
                .map(ToString::to_string)
                .collect::<std::collections::BTreeSet<_>>()
        })
        .unwrap_or_default();

    let properties = obj
        .entry("properties")
        .or_insert_with(|| serde_json::json!({}));

    let Some(properties) = properties.as_object_mut() else {
        return;
    };

    let mut property_names = properties.keys().cloned().collect::<Vec<_>>();
    property_names.sort();
    for property_name in &property_names {
        let Some(property) = properties.get_mut(property_name) else {
            continue;
        };
        normalize_openai_strict_schema_value(property);
        if !original_required.contains(property_name) {
            make_schema_nullable(property);
        }
    }

    obj.insert(
        "required".to_string(),
        serde_json::Value::Array(
            property_names
                .into_iter()
                .map(serde_json::Value::String)
                .collect(),
        ),
    );
}

fn is_object_schema(obj: &serde_json::Map<String, serde_json::Value>) -> bool {
    obj.get("properties").is_some()
        || obj
            .get("type")
            .and_then(serde_json::Value::as_str)
            .is_some_and(|ty| ty == "object")
}

fn make_schema_nullable(schema: &mut serde_json::Value) {
    let Some(obj) = schema.as_object_mut() else {
        return;
    };

    match obj.get_mut("type") {
        Some(serde_json::Value::String(ty)) if ty != "null" => {
            let ty = ty.clone();
            obj.insert(
                "type".to_string(),
                serde_json::Value::Array(vec![
                    serde_json::Value::String(ty),
                    serde_json::Value::String("null".to_string()),
                ]),
            );
        }
        Some(serde_json::Value::Array(types)) => {
            if !types.iter().any(|value| value.as_str() == Some("null")) {
                types.push(serde_json::Value::String("null".to_string()));
            }
        }
        Some(_) => {}
        None => {
            let existing = serde_json::Value::Object(obj.clone());
            obj.clear();
            obj.insert(
                "anyOf".to_string(),
                serde_json::Value::Array(vec![
                    existing,
                    serde_json::json!({
                        "type": "null"
                    }),
                ]),
            );
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn gate_normalizes_text_format_schema_optional_fields() {
        let mut body = serde_json::json!({
            "text": {
                "format": {
                    "type": "json_schema",
                    "schema": {
                        "type": "object",
                        "properties": {
                            "ok": { "type": "boolean" },
                            "reason": { "type": "string" },
                            "impossible": { "type": "boolean" }
                        },
                        "required": ["ok", "reason"]
                    }
                }
            }
        });

        apply_openai_strict_schema_gate(&mut body);

        let schema = &body["text"]["format"]["schema"];
        assert_eq!(
            schema["required"],
            serde_json::json!(["impossible", "ok", "reason"])
        );
        assert_eq!(
            schema["properties"]["impossible"]["type"],
            serde_json::json!(["boolean", "null"])
        );
        assert_eq!(schema["additionalProperties"], false);
    }

    #[test]
    fn gate_normalizes_tool_parameter_schemas() {
        let mut body = serde_json::json!({
            "tools": [{
                "type": "function",
                "name": "Stop",
                "parameters": {
                    "type": "object",
                    "properties": {
                        "decision": { "type": "string" },
                        "reason": { "type": "string" }
                    },
                    "required": ["decision"]
                }
            }]
        });

        apply_openai_strict_schema_gate(&mut body);

        let parameters = &body["tools"][0]["parameters"];
        assert_eq!(
            parameters["required"],
            serde_json::json!(["decision", "reason"])
        );
        assert_eq!(
            parameters["properties"]["reason"]["type"],
            serde_json::json!(["string", "null"])
        );
        assert_eq!(parameters["additionalProperties"], false);
    }
}
